File: zig/src/tools/eval_content_search.zig
//! eval-content-search — offline content_search parity runner.

const std = @import("std");
const zeroclaw = @import("zeroclaw");
const agent_tools = zeroclaw.agent_tools;

const EvalError = error{InvalidScenario};

extern "c" fn setenv(name: [*:0]const u8, value: [*:0]const u8, overwrite: c_int) c_int;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const input = try std.io.getStdIn().readToEndAlloc(allocator, 16 * 1024 * 1024);
    defer allocator.free(input);

    const stdout = std.io.getStdOut().writer();
    var lines = std.mem.splitScalar(u8, input, '\n');
    while (lines.next()) |raw_line| {
        const line = std.mem.trim(u8, raw_line, " \t\r\n");
        if (line.len == 0) continue;
        try runOp(allocator, line, stdout);
    }
}

fn runOp(allocator: std.mem.Allocator, line: []const u8, writer: anytype) !void {
    var parsed = try std.json.parseFromSlice(std.json.Value, allocator, line, .{});
    defer parsed.deinit();

    const op = getString(parsed.value, "op") orelse return EvalError.InvalidScenario;
    if (!std.mem.eql(u8, op, "execute")) return EvalError.InvalidScenario;
    const tool_name = getString(parsed.value, "tool") orelse return EvalError.InvalidScenario;
    if (!std.mem.eql(u8, tool_name, "content_search")) return EvalError.InvalidScenario;
    const setup = getField(parsed.value, "setup") orelse return EvalError.InvalidScenario;
    const args = getField(parsed.value, "args") orelse return EvalError.InvalidScenario;

    try applySetup(allocator, setup);
    const workspace = getString(setup, "workspace") orelse getString(setup, "home") orelse return EvalError.InvalidScenario;
    try std.posix.chdir(workspace);

    var tool_impl = agent_tools.ContentSearchTool.init(allocator);
    defer tool_impl.deinit(allocator);
    var result = try tool_impl.tool().execute(allocator, args);
    defer result.deinit(allocator);
    try writeExecuteResult(writer, tool_name, result);
}

fn applySetup(allocator: std.mem.Allocator, setup: std.json.Value) !void {
    if (getString(setup, "workspace")) |workspace| {
        try std.fs.cwd().makePath(workspace);
    }

    if (getString(setup, "home")) |home| {
        try std.fs.cwd().makePath(home);
        const key = try allocator.dupeZ(u8, "HOME");
        defer allocator.free(key);
        const value = try allocator.dupeZ(u8, home);
        defer allocator.free(value);
        if (setenv(key.ptr, value.ptr, 1) != 0) return error.Unexpected;
    }

    if (getField(setup, "files")) |files| {
        if (files != .array) return EvalError.InvalidScenario;
        for (files.array.items) |file| {
            if (file != .object) return EvalError.InvalidScenario;
            const path = getString(file, "path") orelse return EvalError.InvalidScenario;
            const content = getString(file, "content") orelse return EvalError.InvalidScenario;
            if (std.fs.path.dirname(path)) |parent| try std.fs.cwd().makePath(parent);
            try std.fs.cwd().writeFile(.{ .sub_path = path, .data = content });
        }
    }
}

fn writeExecuteResult(writer: anytype, tool_name: []const u8, result: agent_tools.ToolResult) !void {
    try writer.writeAll("{\"op\":\"execute\",\"tool\":");
    try std.json.stringify(tool_name, .{}, writer);
    try writer.writeAll(",\"result\":{\"success\":");
    try writer.writeAll(if (result.success) "true" else "false");
    try writer.writeAll(",\"output\":");
    try std.json.stringify(result.output, .{}, writer);
    if (result.error_msg) |msg| {
        try writer.writeAll(",\"error\":");
        try std.json.stringify(msg, .{}, writer);
    }
    try writer.writeAll("}}\n");
}

fn getField(value: std.json.Value, key: []const u8) ?std.json.Value {
    if (value != .object) return null;
    return value.object.get(key);
}

fn getString(value: std.json.Value, key: []const u8) ?[]const u8 {
    const inner = getField(value, key) orelse return null;
    if (inner != .string) return null;
    return inner.string;
}

File: eval-tools/src/bin/eval-content-search.rs
//! eval-content-search — offline content_search parity runner.

use std::io::{self, Read, Write};
use std::path::{Path, PathBuf};
use std::sync::Arc;

use anyhow::{Context, Result};
use serde_json::{Map, Value};
use zeroclaw_api::tool::{Tool, ToolResult};
use zeroclaw_config::policy::SecurityPolicy;
use zeroclaw_tools::content_search::ContentSearchTool;

#[tokio::main(flavor = "current_thread")]
async fn main() -> Result<()> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;
    let mut output = Vec::new();

    for raw_line in input.lines() {
        let line = raw_line.trim();
        if line.is_empty() {
            continue;
        }
        let value: Value = serde_json::from_str(line).context("invalid scenario JSON")?;
        let op = required_string(&value, "op")?;
        if op != "execute" {
            anyhow::bail!("unknown content_search op: {op}");
        }

        let tool = required_string(&value, "tool")?;
        if tool != "content_search" {
            anyhow::bail!("unknown content search tool: {tool}");
        }

        let setup = required_field(&value, "setup")?;
        apply_setup(setup)?;
        let workspace = workspace_path(setup)?;
        let security = Arc::new(security_policy(setup, workspace));
        let args = required_field(&value, "args")?.clone();
        let result = execute_tool(security, args).await;
        write_execute_result(&mut output, tool, result)?;
    }

    io::stdout().lock().write_all(&output)?;
    Ok(())
}

async fn execute_tool(security: Arc<SecurityPolicy>, args: Value) -> ToolResult {
    match ContentSearchTool::new_with_backend(security, false)
        .execute(args)
        .await
    {
        Ok(result) => result,
        Err(err) => ToolResult {
            success: false,
            output: String::new(),
            error: Some(err.to_string()),
        },
    }
}

fn security_policy(setup: &Value, workspace: PathBuf) -> SecurityPolicy {
    let security = setup.get("security");
    let allow_absolute_under_root =
        optional_bool(security, "allow_absolute_under_root").unwrap_or(false);
    let allow_resolved_outside_workspace =
        optional_bool(security, "allow_resolved_outside_workspace").unwrap_or(false);
    let rate_limited = optional_bool(security, "rate_limited").unwrap_or(false);

    let mut forbidden_paths = SecurityPolicy::default().forbidden_paths;
    if let Some(extra_blocked_paths) = optional_string_array(security, "extra_blocked_paths") {
        forbidden_paths.extend(extra_blocked_paths);
    }

    SecurityPolicy {
        allowed_roots: if allow_absolute_under_root {
            vec![workspace.clone()]
        } else {
            Vec::new()
        },
        workspace_dir: workspace,
        workspace_only: !allow_resolved_outside_workspace,
        forbidden_paths,
        max_actions_per_hour: if rate_limited { 0 } else { 1000 },
        ..SecurityPolicy::default()
    }
}

fn apply_setup(setup: &Value) -> Result<()> {
    if let Some(workspace) = optional_string(setup, "workspace") {
        std::fs::create_dir_all(workspace)?;
    }

    if let Some(home) = optional_string(setup, "home") {
        std::fs::create_dir_all(home)?;
        std::env::set_var("HOME", home);
    }

    if let Some(files) = setup.get("files") {
        let files = files.as_array().context("setup files must be an array")?;
        for file in files {
            let path = required_string(file, "path")?;
            let content = required_string(file, "content")?;
            if let Some(parent) = Path::new(path).parent() {
                std::fs::create_dir_all(parent)?;
            }
            std::fs::write(path, content)?;
        }
    }

    Ok(())
}

fn workspace_path(setup: &Value) -> Result<PathBuf> {
    optional_string(setup, "workspace")
        .map(PathBuf::from)
        .or_else(|| optional_string(setup, "home").map(PathBuf::from))
        .context("setup workspace missing")
}

fn write_execute_result(output: &mut Vec<u8>, tool: &str, result: ToolResult) -> Result<()> {
    let mut result_obj = Map::new();
    result_obj.insert("success".to_string(), Value::Bool(result.success));
    result_obj.insert("output".to_string(), Value::String(result.output));
    if let Some(error) = result.error {
        result_obj.insert("error".to_string(), Value::String(error));
    }

    let value = serde_json::json!({
        "op": "execute",
        "tool": tool,
        "result": Value::Object(result_obj),
    });
    serde_json::to_writer(&mut *output, &value)?;
    output.push(b'\n');
    Ok(())
}

fn required_field<'a>(value: &'a Value, key: &str) -> Result<&'a Value> {
    value.get(key).with_context(|| format!("{key} missing"))
}

fn required_string<'a>(value: &'a Value, key: &str) -> Result<&'a str> {
    value
        .get(key)
        .and_then(Value::as_str)
        .with_context(|| format!("{key} missing"))
}

fn optional_string<'a>(value: &'a Value, key: &str) -> Option<&'a str> {
    value.get(key).and_then(Value::as_str)
}

fn optional_bool(value: Option<&Value>, key: &str) -> Option<bool> {
    value?.get(key).and_then(Value::as_bool)
}

fn optional_string_array(value: Option<&Value>, key: &str) -> Option<Vec<String>> {
    let items = value?.get(key)?.as_array()?;
    let mut output = Vec::with_capacity(items.len());
    for item in items {
        output.push(item.as_str()?.to_string());
    }
    Some(output)
}
